core/src/main/java/org/testcontainers/dockerclient/DockerClientProviderStrategy.java

```diff
@@ -49,45 +61,51 @@
     public static DockerClientProviderStrategy getFirstValidStrategy(List<DockerClientProviderStrategy> strategies) {
         List<String> configurationFailures = new ArrayList<>();
 
-        for (DockerClientProviderStrategy strategy : strategies) {
-            try {
-                strategy.test();
-                LOGGER.info("Looking for Docker environment. Tried {}", strategy.getDescription());
-                return strategy;
-            } catch (Exception | ExceptionInInitializerError | NoClassDefFoundError e) {
-                @Nullable String throwableMessage = e.getMessage();
-                @SuppressWarnings("ThrowableResultOfMethodCallIgnored")
-                Throwable rootCause = Throwables.getRootCause(e);
-                @Nullable String rootCauseMessage = rootCause.getMessage();
-
-                String failureDescription;
-                if (throwableMessage != null && throwableMessage.equals(rootCauseMessage)) {
-                    failureDescription = String.format("%s: failed with exception %s (%s)",
-                            strategy.getClass().getSimpleName(),
-                            e.getClass().getSimpleName(),
-                            throwableMessage);
-                } else {
-                    failureDescription = String.format("%s: failed with exception %s (%s). Root cause %s (%s)",
-                            strategy.getClass().getSimpleName(),
-                            e.getClass().getSimpleName(),
-                            throwableMessage,
-                            rootCause.getClass().getSimpleName(),
-                            rootCauseMessage
-                    );
-                }
-                configurationFailures.add(failureDescription);
-
-                LOGGER.debug(failureDescription);
-            }
-        }
-
-        LOGGER.error("Could not find a valid Docker environment. Please check configuration. Attempted configurations were:");
-        for (String failureMessage : configurationFailures) {
-            LOGGER.error("    " + failureMessage);
-        }
-        LOGGER.error("As no valid configuration was found, execution cannot continue");
-
-        throw new IllegalStateException("Could not find a valid Docker environment. Please see logs and check configuration");
+        return strategies.stream()
+                .filter(DockerClientProviderStrategy::isApplicable)
+                .sorted(Comparator.comparing(DockerClientProviderStrategy::getPriority).reversed())
```

Comment:
My only slight concern with determining the order at run time is that it means there's no one place you can look in the code to see the sequence. Now people will have to look at the implementations of all the strategies to figure out what order they'll run in.

Having said that, this does allow us to do dynamic prioritisation. So, e.g. as I mentioned on Slack, caching last known good config somewhere. Is that the kind of thing you had in mind too?